src/Kernel32/storebanned/Kernel32+DeviceIOControlOverlapped.cs

```diff
@@ -13,23 +13,30 @@ namespace PInvoke
     /// </content>
     public partial class Kernel32
     {
-        private class DeviceIOControlOverlapped : Overlapped
+        private class DeviceIOControlOverlapped<TInput, TOutput> : Overlapped
+            where TInput : struct
+            where TOutput : struct
         {
-            private readonly Memory<byte> input;
-            private readonly Memory<byte> output;
+            private readonly Memory<TInput> input;
+            private readonly Memory<TOutput> output;
+
+            /// <summary>
+            /// The source for completing the <see cref="Completion"/> property.
+            /// </summary>
+            private readonly TaskCompletionSource<uint> completion = new TaskCompletionSource<uint>();
 
             private unsafe NativeOverlapped* native;
 
             /// <summary>
-            /// Initializes a new instance of the <see cref="DeviceIOControlOverlapped"/> class.
+            /// Initializes a new instance of the <see cref="DeviceIOControlOverlapped{TInput, TOutput}"/> class.
             /// </summary>
             /// <param name="input">
             /// The input buffer.
             /// </param>
             /// <param name="output">
             /// The output buffer.
             /// </param>
-            public DeviceIOControlOverlapped(Memory<byte> input, Memory<byte> output)
+            internal DeviceIOControlOverlapped(Memory<TInput> input, Memory<TOutput> output)
```

Comment:
Out of curiosity, since this constructor (and the properties below) are defined on a `class` which is `private` to `Kernel32`, is there any difference between it as `public` or `internal`?

Reply:
There are subtle differences in Reflection, but the main reason is because `public` triggers "public API review" mode in my mind as public APIs are very hard to change later, and these are in fact *not* public APIs. The C# `public` keyword in `private` or `internal` types are in fact not public at all (unless they also implement an interface). So it's a misleading keyword to use, and makes code reviews more expensive. So I prefer `internal` for members of non-public types instead of `public`.